### PR TITLE
[MXNET-246] operators for Synchronized BatchNorm

### PR DESCRIPTION
**Pull request by zhanghang1989**
## Description ##
Backend operators for Synchronized Batch Norm. Design idea as in http://hangzh.com/SynchronizeBN/

Special thanks to Haibin @eric-haibin-lin 

1. AllReduce 
2. Decouple Batch Norm
3. SumSquare

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] AllReduce, tests, (and when applicable, API doc)
- [ ] Decouple Batch Norm, tests, (and when applicable, API doc)
- [ ] SumSquare,  tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/imperative/imperative_utils.h [hunk 1]
```diff
@@ -164,7 +164,12 @@ inline void SetShapeType(const Context& ctx,
     NDArrayStorageType storage_type = static_cast<NDArrayStorageType>(out_storage_types[i]);
     if (outputs[i]->is_none()) {
       if (storage_type == kDefaultStorage) {
-        *outputs[i] = NDArray(out_shapes[i], ctx, true, out_types[i]);
+        if (outputs.size() == inputs.size()) {
```
Comment (piiswrong) [inline src/imperative/imperative_utils.h]: ???

Comment (zhanghang1989) [reply src/imperative/imperative_utils.h]: This is for AllReduce operator, the outputs should be in the same device as the corresponding input.

**Comment by zhanghang1989** [comment]
Only all-reduce operation need cross-device communication. The other operations should be calculated individually at different device. That's why I separate those operations to different "operators".

File: src/imperative/imperative_utils.h (hunk 1 above)
Comment (piiswrong) [reply src/imperative/imperative_utils.h]: This is too hacky

**Comment by piiswrong** [comment]
why not combine them into the same operator? It would save memory and be much more efficient

File: src/imperative/imperative_utils.h (hunk 1 above)
Comment (zhanghang1989) [reply src/imperative/imperative_utils.h]: I aggree. I am not familiar with very backend code. Could you give suggestions?

**Comment by chinakook** [comment]
There are someone in TuSimple shared idea of this. It may help you.
https://zhuanlan.zhihu.com/p/27069202

**Comment by zhanghang1989** [comment]
Thanks @chinakook  Actually, we already implemented synchronized batch normalization internally. Just thinking about releasing a proper API for the user.

**Comment by zhanghang1989** [comment]
MXNet Gluon Cross-GPU Batch Norm is Implemented here https://github.com/zhanghang1989/MXNet-Gluon-SyncBN . Feel free to try and leave comments.

**Comment by zhanghang1989** [comment]
Closing this in favor of the new PR https://github.com/apache/incubator-mxnet/pull/11502
